Generic num types (by Danny Wolf)

alphaville opened the pull request:
Original PR: #9 (thanks to @wolfd)

## Main changes

- Introduced the trait `LbfgsPrecision` which defines certain tolerances.
- The struct `Lbfgs` uses a generic (defaults to `f64` for backward compatibility); this was done to support `f32`
- Thorough unit testing (both `f64` and `f32` implementations)
- GitHub Actions now runs the unit tests

## Version

New version [`0.3.0-alpha.1`](https://crates.io/crates/lbfgs/0.3.0-alpha.1). This has been tested with OpEn and it works, so after we merge we can release version `0.3.0`.

wolfd committed:
Make lbfgs generic over float type

Topic: lbfgs-rs
Relative: websockets

alphaville committed:
more tests + Precision trair

alphaville committed:
unit test correctneess_reset

alphaville committed:
unit test correctneess_buff_overfull

alphaville committed:
unit test lbfgs_buffer_storage

alphaville committed:
GA unit testing

alphaville committed:
correctneess_buff_2 unit test

alphaville committed:
documentation

korken89 reviewed (commented):
Are we sure the debug asserts should be full blown asserts? I'm sort of remembering the was some reason behind it?

alphaville committed:
assert to debug_assert

alphaville commented:
> Are we sure the debug asserts should be full blown asserts? I'm sort of remembering the was some reason behind it?

We had used `debug_assert!` so that when compiling with `--release` we don't waste time checking conditions that should be satisfied. I changed them to `assert!` so that `cargo test --release` would work, but I've now changed them back to `debug_assert!`. If we want to run the unit tests with `--release` we can do

```bash
RUSTFLAGS="-C debug-assertions" cargo build --release
```

alphaville committed:
update readme (docs)

alphaville committed:
update LICENCE-MIT

alphaville committed:
update LICENSE-APACHE (names)

alphaville committed:
README: fix GA badge

alphaville committed:
update README: further details

alphaville committed:
typo in lib.rs docs

alphaville committed:
remove .travis.yml (not used)

korken89 commented:
Looks good 👍

alphaville committed:
version 0.3.0
